lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1172,80 +1173,63 @@ class TypeAccessScopeChecker : private TypeWalker {
   using TypeAccessScopeCacheMap = TypeChecker::TypeAccessScopeCacheMap;
   TypeAccessScopeCacheMap &Cache;
   const SourceFile *File;
-  SmallVector<intptr_t, 8> RawScopeStack;
-
-  static constexpr const intptr_t INVALID = -1;
+  SmallVector<Optional<AccessScope>, 8> RawScopeStack;
 
   explicit TypeAccessScopeChecker(TypeAccessScopeCacheMap &cache,
                                   const SourceFile *file)
       : Cache(cache), File(file) {
     // Always have something on the stack.
-    RawScopeStack.push_back((intptr_t)INVALID);
+    RawScopeStack.push_back(None);
   }
 
   bool shouldVisitOriginalSubstitutedType() override { return true; }
 
-  static intptr_t intersectAccess(const intptr_t first, const intptr_t second) {
-    if (first == INVALID || second == INVALID)
-      return INVALID;
-    if (!first)
-      return second;
-    if (!second)
-      return first;
-    if (first == second)
-      return first;
-
-    auto firstDC = reinterpret_cast<const DeclContext *>(first);
-    auto secondDC = reinterpret_cast<const DeclContext *>(second);
-    if (firstDC->isChildContextOf(secondDC))
-      return first;
-    if (secondDC->isChildContextOf(firstDC))
-      return second;
-    return INVALID;
-  }
-
   Action walkToTypePre(Type ty) override {
     // Assume failure until we post-visit this node.
     // This will be correct as long as we don't ever have self-referential
     // Types.
     auto cached = Cache.find(ty);
     if (cached != Cache.end()) {
-      auto opaqueCached = reinterpret_cast<intptr_t>(cached->second);
-      RawScopeStack.back() = intersectAccess(RawScopeStack.back(),opaqueCached);
+      auto last = &RawScopeStack.back();
+      if (last->hasValue())
+        *last = last->getValue().intersectWith(cached->second);
       return Action::SkipChildren;
     }
 
-    const DeclContext *DC;
+    AccessScope AS;
     if (auto alias = dyn_cast<NameAliasType>(ty.getPointer()))
-      DC = alias->getDecl()->getFormalAccessScope(File);
+      AS = alias->getDecl()->getFormalAccessScope(File);
     else if (auto nominal = ty->getAnyNominal())
-      DC = nominal->getFormalAccessScope(File);
-    else
-      DC = nullptr;
-    RawScopeStack.push_back(reinterpret_cast<intptr_t>(DC));
+      AS = nominal->getFormalAccessScope(File);
+    RawScopeStack.push_back(AS);
 
     return Action::Continue;
   }
 
   Action walkToTypePost(Type ty) override {
     auto last = RawScopeStack.pop_back_val();
-    if (last != INVALID)
-      Cache[ty] = reinterpret_cast<const DeclContext *>(last);
-    RawScopeStack.back() = intersectAccess(RawScopeStack.back(), last);
+    if (last.hasValue()) {
+      Cache[ty] = *last;
+
+      auto prev = &RawScopeStack.back();
+      if (prev->hasValue())
+        *prev = prev->getValue().intersectWith(*last);
+    }
+
     return Action::Continue;
   }
 
 public:
-  static Optional<const DeclContext *>
+  static const AccessScope *
```

Comment:
I don't like the idea of handing out pointers to the DenseMap internal state. I think a value-type return is still safer here.